lib/datadog/appsec/processor.rb

```diff
@@ -0,0 +1,116 @@
+# typed: ignore
+
+require 'datadog/appsec/assets'
+
+module Datadog
+  module AppSec
+    # Processor integrates libddwaf into datadog/appsec
+    class Processor
+      # Interface object to check using case .. when
+      module IOLike
+        def read; end
+        def rewind; end
+
+        def self.===(other)
+          instance_methods.all? { |meth| other.respond_to?(meth) }
+        end
+      end
+
+      def initialize
+        @ruleset = nil
+        @handle = nil
+
+        unless load_libddwaf && load_ruleset && create_waf_handle
+          Datadog.logger.warn { 'AppSec is disabled, see logged errors above' }
+        end
+      end
+
+      def ready?
+        !@ruleset.nil? && !@handle.nil?
+      end
+
+      def new_context
+        Datadog::AppSec::WAF::Context.new(@handle)
+      end
+
+      private
+
+      def load_libddwaf
+        Processor.require_libddwaf && Processor.libddwaf_required?
+      end
+
+      def load_ruleset
+        ruleset_setting = Datadog::AppSec.settings.ruleset
+
+        begin
+          @ruleset = case ruleset_setting
+                     when :recommended, :risky, :strict
+                       JSON.parse(Datadog::AppSec::Assets.waf_rules(ruleset_setting))
+                     when String
+                       JSON.parse(File.read(ruleset_setting))
+                     when IOLike
+                       JSON.parse(ruleset_setting.read).tap { ruleset_setting.rewind }
+                     when Hash
+                       ruleset_setting
+                     else
+                       raise ArgumentError, "unsupported value for ruleset setting: #{ruleset_setting.inspect}"
+                     end
+
+          true
+        rescue StandardError => e
+          Datadog.logger.error do
+            "libddwaf ruleset failed to load, ruleset: #{ruleset_setting.inspect} error: #{e.inspect}"
```

Comment:
Minor: I wonder if in this case it'd make sense to print the error before the ruleset, since the ruleset may be a huge multi-line thing. Or perhaps cut off the maximum number of chars printed from the ruleset?

Reply:
Good point, but it's the other way around for the most common case that customers will face, and in the cases of non-files I'd like the full output. I'll think about how to improve on that.